internal/controller/streams.go

```diff
@@ -197,105 +164,225 @@ func (sm *streamManager) BootUpStreamAndReconnect(ctx context.Context, cancel co
+			if err != nil {
+				sm.logger.Warnw("Failed to listen to flows", "error", err)
+				return err
 			}
-		}()
+		}
 	}
 	return nil
 }
 
-// ExponentialStreamConnect will continue to reboot and restart the main operations within the operator if any disconnects or errors occur.
-func ExponentialStreamConnect(ctx context.Context, logger *zap.SugaredLogger, envMap EnvironmentConfig, bufferedGrpcSyncer *BufferedGrpcWriteSyncer) {
+// connectAndStreamCiliumNetworkFlows creates ciliumNetworkFlows client and begins the streaming of network flows.
+func connectAndStreamCiliumNetworkFlows(logger *zap.SugaredLogger, sm *streamManager) error {
+	ciliumCtx, ciliumCancel := context.WithCancel(context.Background())
+	defer ciliumCancel()
+
+	sendCiliumNetworkFlowsStream, err := sm.streamClient.client.SendKubernetesNetworkFlows(ciliumCtx)
+	if err != nil {
+		logger.Errorw("Failed to connect to server", "error", err)
+		return err
+	}
+
+	sm.streamClient.ciliumNetworkFlowsStream = sendCiliumNetworkFlowsStream
+
+	err = sm.StreamCiliumNetworkFlows(ciliumCtx, ciliumCancel, sm.streamClient.ciliumNamespace)
+	if err != nil {
+		logger.Errorw("Failed to bootup and Cilium network flows", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// connectAndStreamResources creates resourceStream client and begins the streaming of resources.
+func connectAndStreamResources(logger *zap.SugaredLogger, sm *streamManager) error {
+	resourceCtx, resourceCancel := context.WithCancel(context.Background())
+	defer resourceCancel()
+
+	SendKubernetesResourcesStream, err := sm.streamClient.client.SendKubernetesResources(resourceCtx)
+	if err != nil {
+		logger.Errorw("Failed to connect to server", "error", err)
+		return err
+	}
+
+	sm.streamClient.resourceStream = SendKubernetesResourcesStream
+
+	err = sm.StreamResources(resourceCtx, resourceCancel)
+	if err != nil {
+		logger.Errorw("Failed to bootup and stream resources", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// connectAndStreamLogs creates sendLogs client and begins the streaming of logs.
+func connectAndStreamLogs(logger *zap.SugaredLogger, sm *streamManager) error {
+	logCtx, logCancel := context.WithCancel(context.Background())
+	defer logCancel()
+
+	SendLogsStream, err := sm.streamClient.client.SendLogs(logCtx)
+	if err != nil {
+		logger.Errorw("Failed to connect to server", "error", err)
+		return err
+	}
+
+	sm.streamClient.logStream = SendLogsStream
+	sm.bufferedGrpcSyncer.UpdateClient(sm.streamClient.logStream, sm.streamClient.conn)
+
+	err = sm.StreamLogs(logCtx, logCancel)
+	if err != nil {
+		logger.Errorw("Failed to bootup and stream logs", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// Generic function to manage any stream with backoff and reconnection logic.
+func manageStream(logger *zap.SugaredLogger, connectAndStream func(*zap.SugaredLogger, *streamManager) error, sm *streamManager, done chan struct{}) {
 	var backoff = 1 * time.Second
```

Comment:
As an alternative to returning after 10 retries, you could make this function loop indefinitely and:
* Decrease the backoff automatically e.g. after some time without any errors, or after a successful connection. Otherwise, if the pod has any temporary connectivity issues, the backoff would remain forever high until the pod is restarted.
* Put a cap on the backoff, so it would not exceed a maximum. E.g. 30s, 1m, whatever.

Just a thought...